src/graph/sampler.cc

```diff
@@ -997,6 +998,124 @@ Subgraph NegEdgeSubgraph(int64_t num_tot_nodes, const Subgraph &pos_subg,
+                            bool exclude_positive) {
+  std::vector<IdArray> adj = pos_subg.graph->GetAdj(false, "coo");
+  IdArray coo = adj[0];
+  int64_t num_pos_edges = coo->shape[0] / 2;
+
+  int64_t chunk_size = neg_sample_size;
+  // If num_pos_edges isn't divisible by chunk_size, the actual number of chunks
+  // is num_chunks + 1 and the last chunk size is last_chunk_size.
+  // Otherwise, the actual number of chunks is num_chunks, the last chunk size
+  // is 0.
+  int64_t num_chunks = num_pos_edges / chunk_size;
+  int64_t last_chunk_size = num_pos_edges - num_chunks * chunk_size;
+
+  // The number of negative edges.
+  int64_t num_neg_edges = neg_sample_size * chunk_size * num_chunks;
+  int64_t num_neg_edges_last_chunk = neg_sample_size * last_chunk_size;
+  int64_t num_all_neg_edges = num_neg_edges + num_neg_edges_last_chunk;
+
+  // We should include the last chunk.
+  if (last_chunk_size > 0)
+    num_chunks++;
+
+  IdArray neg_dst = IdArray::Empty({num_all_neg_edges}, coo->dtype, coo->ctx);
+  IdArray neg_src = IdArray::Empty({num_all_neg_edges}, coo->dtype, coo->ctx);
+  IdArray neg_eid = IdArray::Empty({num_all_neg_edges}, coo->dtype, coo->ctx);
+  IdArray induced_neg_eid = IdArray::Empty({num_all_neg_edges}, coo->dtype, coo->ctx);
+
+  // These are vids in the positive subgraph.
+  const dgl_id_t *dst_data = static_cast<const dgl_id_t *>(coo->data);
+  const dgl_id_t *src_data = static_cast<const dgl_id_t *>(coo->data) + num_pos_edges;
+  const dgl_id_t *induced_vid_data = static_cast<const dgl_id_t *>(pos_subg.induced_vertices->data);
+  const dgl_id_t *induced_eid_data = static_cast<const dgl_id_t *>(pos_subg.induced_edges->data);
+  size_t num_pos_nodes = pos_subg.graph->NumVertices();
+  std::vector<size_t> pos_nodes(induced_vid_data, induced_vid_data + num_pos_nodes);
+
+  dgl_id_t *neg_dst_data = static_cast<dgl_id_t *>(neg_dst->data);
+  dgl_id_t *neg_src_data = static_cast<dgl_id_t *>(neg_src->data);
+  dgl_id_t *neg_eid_data = static_cast<dgl_id_t *>(neg_eid->data);
+  dgl_id_t *induced_neg_eid_data = static_cast<dgl_id_t *>(induced_neg_eid->data);
+
+  bool neg_head = (neg_mode == "head");
+
+  const dgl_id_t *unchanged;
+  dgl_id_t *neg_unchanged;
+  dgl_id_t *neg_changed;
+
+  // corrupt head nodes.
+  if (neg_head) {
+    unchanged = dst_data;
+    neg_unchanged = neg_dst_data;
+    neg_changed = neg_src_data;
+  } else {
+    // corrupt tail nodes.
+    unchanged = src_data;
+    neg_unchanged = neg_src_data;
+    neg_changed = neg_dst_data;
+  }
+
+  // We first sample all negative edges.
+  std::vector<size_t> neg_vids;
+  RandomSample(num_tot_nodes,
+               num_chunks * neg_sample_size,
+               &neg_vids);
+
+  dgl_id_t curr_eid = 0;
+  std::unordered_map<dgl_id_t, dgl_id_t> neg_map;
+  for (int64_t i_chunk = 0; i_chunk < num_chunks; i_chunk++) {
+    // for each chunk.
+    size_t neg_idx = neg_sample_size * chunk_size * i_chunk;
```

Comment:
size_t and int64_t may not be compatible

Reply:
It should be fine, but I'll make the type consistent.